src/storage/indexeddb/storage.rs

```diff
@@ -0,0 +1,439 @@
+fn js2uuid(js: JsValue) -> Result<Uuid> {
+    let json = js.as_string().ok_or_else(invalid)?;
+    let uuid = Uuid::parse_str(json.as_str()).map_err(invalid_err)?;
+    Ok(uuid)
+}
+
+/// Convert a UUID to a JsValue.
+fn uuid2js(uuid: Uuid) -> Result<JsValue> {
+    let string = uuid.to_string();
+    let js = JsValue::from_str(string.as_str());
+    Ok(js)
+}
+
+/// Convert a JsValue to a Task.
+fn js2task(js: JsValue) -> Result<TaskMap> {
+    Ok(serde_wasm_bindgen::from_value(js)?)
+}
+
+/// Convert a Task to a JsValue.
+fn task2js(task: TaskMap) -> Result<JsValue> {
+    Ok(serde_wasm_bindgen::to_value(&task)?)
+}
+
+fn unsynced_is_zero(v: &u8) -> bool {
+    *v == 0
+}
+
+/// Operations are stored with a separate UUID (for IndexedDB indexing) and
+/// a flag indicating whether they have been synced.
+#[derive(Serialize, Deserialize)]
+struct StoredOperation {
+    uuid: Option<Uuid>,
+    operation: Operation,
+
+    // IndexedDB does not support indexing by a boolean value, so
+    // we store a number here instead. Since we want to index un-synced
+    // operations, this field is only present on such entries.
+    #[serde(skip_serializing_if = "unsynced_is_zero")]
+    #[serde(default)]
+    unsynced: u8,
+}
+
+/// Convert a JsValue to a Task.
+fn js2op(js: JsValue) -> Result<Operation> {
+    Ok(serde_wasm_bindgen::from_value::<StoredOperation>(js)?.operation)
+}
+
+/// Convert a Task to a JsValue.
+fn op2js(operation: Operation, unsynced: bool) -> Result<JsValue> {
+    let operation = StoredOperation {
+        uuid: operation.get_uuid(),
+        operation,
+        unsynced: unsynced as u8,
+    };
+    Ok(serde_wasm_bindgen::to_value(&operation)?)
+}
+
+struct InnerTxn(Option<Transaction>);
+
+impl InnerTxn {
+    fn idb_txn(&self) -> Result<&idb::Transaction> {
+        if let Some(transaction) = &self.0 {
+            Ok(transaction)
+        } else {
+            Err(already_committed())
+        }
+    }
+
+    async fn get_next_working_set_number(&self) -> Result<usize> {
+        let working_set = self.idb_txn()?.object_store(schema::WORKING_SET)?;
+        let mut max: usize = 0;
+        let mut maybe_cursor = working_set.open_key_cursor(None, None)?.await?;
+        while let Some(cursor) = maybe_cursor {
+            let i = cursor.key()?.as_f64().ok_or_else(invalid)? as usize;
+            if i > max {
+                max = i;
+            }
+            maybe_cursor = cursor.next(None)?.await?;
+        }
+        Ok(max + 1)
+    }
+}
+
+#[async_trait(?Send)]
+impl WrappedStorageTxn for InnerTxn {
+    async fn get_task(&mut self, uuid: Uuid) -> Result<Option<TaskMap>> {
+        let tasks = self.idb_txn()?.object_store(schema::TASKS)?;
+        if let Some(task) = tasks.get(Query::Key(uuid2js(uuid)?))?.await? {
+            Ok(Some(js2task(task)?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn get_pending_tasks(&mut self) -> Result<Vec<(Uuid, TaskMap)>> {
+        todo!()
```

Comment:
I didn't know about this macro. Handy! Did you mean to commit it though?

Reply:
Oh! There are no tests for that method! Yikes!